### PR TITLE
Refactor user to account

### PR DESCRIPTION
**Pull request by bgavrilMS**
This is a partial fix, please do not merge it. I'd like to get an opinion on the current approach and how the new API looks like. 

This contains: 

- functional code
- existing unit test pass
- some public comments

Remaining: 

- more unit tests
- another pass through internal code for references to "user" which should be renamed to "account" 
- E2E manual testing

### CODE REVIEW

File: msal/src/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 1]
```diff
@@ -66,9 +66,9 @@ public abstract class ClientApplicationBase
             Authority = authorityInstance.CanonicalAuthority;
             RedirectUri = redirectUri;
             ValidateAuthority = validateAuthority;
-            if (UserTokenCache != null)
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/ClientApplicationBase.cs]: I'm not sure about this renaming, @bgavrilMS

**Review by jmprieur — APPROVED** [review]
LGTM. I've asked a question

File: msal/src/Microsoft.Identity.Client/Account.cs [hunk 2]
```diff
@@ -25,48 +25,37 @@
...
-        public string DisplayableId { get; internal set; }
+        public string Username { get; internal set; }
 
-        /// <summary>
-        /// Gets given name of the user if provided by the service. If not, the value is null.
-        /// </summary>
         public string Environment { get; internal set; }
 
-        public string Identifier { get; internal set; }
+        public MsalAccountId HomeAccountId { get; internal set; }
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/Account.cs]: Could we have AccountId and not MsalAccountId (none of the MSAL.NET classes start with Msal, except for exceptions)

File: msal/src/Microsoft.Identity.Client/Account.cs [hunk 3]
```diff
@@ -25,48 +25,37 @@
...
 
 namespace Microsoft.Identity.Client
 {
     /// <summary>
-    /// Contains information of a single user. This information is used for token cache lookup and enforcing the user session on STS authorize endpont.
+    /// Contains information of a single account. A user can be present in multiple directorie and thus have multiple accounts.
+    /// This information is used for token cache lookup and enforcing the user session on STS authorize endpont.
     /// </summary>
-    internal sealed class User: IUser
+    internal sealed class Account: IAccount
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/Account.cs:38]: Could we please add an override of ToString() (returning the identifier)

File: msal/src/Microsoft.Identity.Client/MsalAccountId.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client
+{
+    /// <summary>
+    /// An identifier for an account in a specific tenant
+    /// </summary>
+    public class MsalAccountId
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/MsalAccountId.cs]: Can we please rename to AccountId (see other comment)
and add a ToString() override returning the Identifier

**Review by jmprieur — APPROVED** [review]
LGTM. 
I've proposed a couple of suggestions.

File: msal/src/Microsoft.Identity.Client/Account.cs (hunk 3 above)
Comment (bgavrilMS) [reply msal/src/Microsoft.Identity.Client/Account.cs:38]: Done

File: msal/samples/desktop/SampleApp/MainForm.cs [hunk 5]
```diff
@@ -36,14 +36,14 @@ namespace SampleApp
     public partial class MainForm : Form
     {
         private readonly MsalAuthHelper _msalHelper = new MsalAuthHelper("11744750-bfe5-4818-a1c0-655455f68fa7");
-        private IUser user = null;
+        private IAccount user = null;
```
Comment (henrik-me) [inline msal/samples/desktop/SampleApp/MainForm.cs:39]: >user [](start = 25, length = 4)

nit: variable should probably be account

File: msal/samples/desktop/SampleApp/MsalAuthHelper.cs [hunk 6]
```diff
@@ -63,7 +63,7 @@ public async Task<IUser> SignInAsync()
             return null;
         }
 
-        public async Task<string> GetTokenForCurrentUserAsync(IEnumerable<string> scopes, IUser user)
+        public async Task<string> GetTokenForCurrentUserAsync(IEnumerable<string> scopes, IAccount user)
```
Comment (henrik-me) [inline msal/samples/desktop/SampleApp/MsalAuthHelper.cs:66]: account?

**Review by henrik-me — APPROVED** [review]
:shipit:

**Comment by bgavrilMS** [comment]
I addressed all the PR comments. Will rebase and merge manually

**Comment by henrik-me** [comment]
not everything :)

---
In reply to: [410955591](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1169#issuecomment-410955591) [](ancestors = 410955591)

**Pull request closed without merging**